elastic/kibana
# Migrate defaultIndex attribute for config saved object

Resolves #133241

### Overview

This PR adds a new mechanism to apply custom _transform_ functions to the `config` saved object for UI settings. These transform functions are complementary to saved object migrations.

The `config` saved objects are versioned and isolated to each space.
When a client attempts to fetch or create a `config` saved object, Kibana first checks if there's already an old one that needs to be upgraded. This PR takes advantage of the existing functionality to simply call any necessary transform functions during that process. This way, we can ensure that transforms will be applied to each `config` after upgrading, but we don't need to add a complicated "on start" loop to search and update saved objects.

### Testing

To manually test this end-to-end to verify the migration is applied:

1. Load the test data and scripts that are attached to #125960
   - This will create a bunch of test data, but we are primarily interested in the index pattern `ff959d40-b880-11e8-a6d9-e546fe2bba5f` in the Default space, along with a legacy URL alias `redirect-with-toast` that points to the aforementioned ID
2. Log in with `superduperuser`/`changeme` which was created by the script and has access to modify system indices
3. Using Dev Tools, create an old config object that will be upgraded, which uses the legacy URL alias ID:
   ```
   POST .kibana/_doc/config:8.2.0
   {
     "config": {
       "buildNum": 123,
       "defaultIndex": "redirect-with-toast"
     },
     "type": "config",
     "references": [],
     "migrationVersion": {
       "config": "8.1.0"
     },
     "coreMigrationVersion": "8.2.0",
     "updated_at": "2022-06-07T13:21:34.759Z"
   }
   ```
4. Using Dev Tools, delete the existing config object (for 8.4.0):
   ```
   DELETE .kibana/_doc/config:8.4.0
   ```
5. Refresh the page to fetch the config object again, triggering the upgrade
6. Using Dev Tools, search for the config objects and verify that it was upgraded correctly:
   ```
   GET .kibana/_search
   {
     "query": {
       "term": {
         "type": {
           "value": "config"
         }
       }
     }
   }
   ```
   The result should look like this:
   ```json
   "hits": [
     {
       "_index": ".kibana_8.4.0_001",
       "_id": "config:8.2.0",
       "_score": 3.6888795,
       "_source": {
         "config": {
           "buildNum": 123,
           "defaultIndex": "redirect-with-toast"
         },
         "type": "config",
         "references": [],
         "migrationVersion": {
           "config": "8.1.0"
         },
         "coreMigrationVersion": "8.4.0",
         "updated_at": "2022-06-07T13:21:34.759Z"
       }
     },
     {
       "_index": ".kibana_8.4.0_001",
       "_id": "config:8.4.0",
       "_score": 3.6888795,
       "_source": {
         "config": {
           "buildNum": 9007199254740991,
           "isDefaultIndexMigrated": true,
           "defaultIndex": "ff959d40-b880-11e8-a6d9-e546fe2bba5f"
         },
         "type": "config",
         "references": [],
         "migrationVersion": {
           "config": "8.1.0"
         },
         "coreMigrationVersion": "8.4.0",
         "updated_at": "2022-06-07T13:32:22.612Z"
       }
     }
   ]
   ```

-----

**Release note:** The advanced setting for `defaultIndex` could be broken (not pointing to the correct index pattern) if it was set in a custom space before upgrading to the 8.0 release. The setting is now migrated correctly the first time that space is accessed by any user that has `All` access to the Advanced Settings feature.
